apps/aecore/src/aec_miner.erl

```diff
@@ -0,0 +1,199 @@
+%%------------------------------------------------------------------------------
+-spec get_balance() -> integer() | {'error', term()}.
+get_balance() ->
+    {ok, Pubkey} = aec_keys:pubkey(),
+    {ok, LastBlock} = aec_chain:top(),
+    Trees = aec_blocks:trees(LastBlock),
+    AccountsTree = aec_trees:accounts(Trees),
+    case aec_accounts:get(Pubkey, AccountsTree) of
+        {ok, #account{balance = B}} ->
+            B;
+        _ ->
+            {error, account_not_found}
+    end.
+
+%%%===================================================================
+%%% gen_statem callbacks
+%%%===================================================================
+
+%%--------------------------------------------------------------------
+%% @private
+%% @doc
+%% Whenever a gen_statem is started using gen_statem:start/[3,4] or
+%% gen_statem:start_link/[3,4], this function is called by the new
+%% process to initialize.
+%%
+%% @spec init(Args) -> {ok, StateName, State} |
+%%                     {ok, StateName, State, Timeout} |
+%%                     ignore |
+%%                     {stop, StopReason}
+%% @end
+%%--------------------------------------------------------------------
+init(_) ->
+    {ok, _} = aec_chain:start_link(aec_block_genesis:genesis_block()),
+    {ok, idle, #state{}}.
+
+%%--------------------------------------------------------------------
+%% @private
+%% @doc
+%% There should be one instance of this function for each possible
+%% state name. Whenever a gen_statem receives an event sent using
+%% gen_statem:send_event/2, the instance of this function with the same
+%% name as the current state name StateName is called to handle
+%% the event. It is also called if a timeout occurs.
+%%
+%% @spec state_name(Type, Event, State) ->
+%%                   {next_state, NextStateName, NextState} |
+%%                   {next_state, NextStateName, NextState, [Actions]}
+%% @end
+%%--------------------------------------------------------------------
+idle({call, From}, start, State) ->
+    case aec_keys:pubkey() of
+        {ok, _Pubkey} ->
+            gen_statem:cast(?SERVER, mine),
+            {next_state, running, State, [{reply, From, ok}]};
+        {error, _} = Error ->
+            lager:error("Cannot start mining: ~p", [Error]),
+            {next_state, idle, State, [{reply, From, Error}]}
+    end;
+idle({call, From}, stop, State) ->
+    {next_state, idle, State, [{reply, From, {error, not_started}}]};
+idle({call, From}, _Msg, State) ->
+    {next_state, idle, State, [{reply, From, {error, not_supported}}]};
+idle(_Type, _Msg, State) ->
+    {next_state, idle, State}.
+
+running({call, From}, start, State) ->
+    {next_state, running, State, [{reply, From, {error, already_started}}]};
+running({call, From}, stop, State) ->
+    {ok, TopBlock} = aec_chain:top(),
+    Height = aec_blocks:height(TopBlock),
+    Bal = get_balance(),
+    lager:info("Mining finished at height ~p, balance ~p", [Height, Bal]),
```

Comment:
Please remove these 4 lines (getting height of top block, getting balance). A client interested in that can do it itself - without requiring the context of the miner.

(If you want to (I would prefer not), you can put those 4 lines in the `aec_miner:stop_mining` function i.e. in the context of the caller process - not the miner.)